### PR TITLE
Fix test workflow for pull requests

### PR DESCRIPTION
**Pull request by jdhameli**
Use pull_request_target, but allow the workflow to run only after the PR has a 'safe to test' label.

### CODE REVIEW

**Comment by mcafaro** [comment]
A few suggestions:

- Break this particular workflow out into its own workflow file (like .github/workflows/test-pull-request.yml). It does require repeating yourself a little bit but given the security implications of `pull_request_target` + `refs/pull/${{ github.event.number }}/merge` it is best to isolate this workflow and have good visibility and control over when/how it changes.
- Use the [`permissions`](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs) key on this workflow to restrict the `GITHUB_TOKEN` to read-only (or none) capabilities. By default, workflows that are triggered by `pull_request_target` have a `GITHUB_TOKEN` with read/write repository permissions which may be exploited when executing arbitrary code from a pull request.
- Immediately after the checkout, verify that `HEAD^` (e.g. `git rev-parse HEAD^`) matches `github.event.pull_request.head.sha`. If these two SHAs do not match, stop the workflow. As I mentioned, there is a window of time after when you label the pull and before the job runs that the user could submit a malicious change. You want to verify that has not occurred.
- Consider setting up a [`CODEOWNERS`](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) file pointing at the workflow file and setting up your repo to require approval from the code owners to change it.

Also, please read https://securitylab.github.com/research/github-actions-preventing-pwn-requests/ and understand the **TLDR:** "Combining pull_request_target workflow trigger with an explicit checkout of an untrusted PR is a dangerous practice that may lead to repository compromise."

File: .github/workflows/test-pull-request.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        version: [R2021b, R2023a]
+        os: [windows-latest, ubuntu-latest]
+    env:
+        MATHWORKS_SKIP_ACTIVATION: true
+        MLM_WEB_LICENSE: true
+        MLM_WEB_ID: ${{secrets.MLM_WEB_ID}}
+        MLM_WEB_USER_CRED: ${{secrets.MLM_WEB_USER_CRED}}
```
Comment (mcafaro) [inline .github/workflows/test-pull-request.yml:27]: Try to limit the scope of these env vars as much as possible. `MATHWORKS_SKIP_ACTIVATION` should only be necessary for `setup-matlab`:

```
      - name: Set up MATLAB
         uses: matlab-actions/setup-matlab@v1.2.3
         with:
           release: ${{ matrix.version }}
         env:
           MATHWORKS_SKIP_ACTIVATION: true
```

The MLM env vars should only be necessary for actions that run MATLAB (just `npm run tests`?).

File: .github/workflows/test-pull-request.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# Run Extension tests
```
Comment (mcafaro) [inline .github/workflows/test-pull-request.yml:1]:
```suggestion
# Run tests on a pull request after it has been labeled 'safe to test'
```

**Review by mcafaro — COMMENTED** [review]
Looking good!

File: .github/workflows/test-pull-request.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          submodules: recursive
+          ref: "refs/pull/${{ github.event.number }}/merge"
+          persist-credentials: false
+      - name: Verify Head
+        shell: bash
+        run: |
+          if [ "$(git ls-remote origin ${{ github.event.number }}/head | awk '{print $1}')" != "${{ github.event.pull_request.head.sha }}" ]; then
```
Comment (mcafaro) [inline .github/workflows/test-pull-request.yml:33]: Interesting. You might want to try confirming this works as expected by attempting to submit a change to a pull request right after you label it 'safe to test' but before the checkout action occurs.

**Review by mcafaro — APPROVED** [review]
:shipit:

File: .github/workflows/test-pull-request.yml (hunk 3 above)
Comment (jdhameli) [reply .github/workflows/test-pull-request.yml:33]: Thanks Mark! I forked my fork to test this and can confirm that the workflow now runs successfully on pull requests and fails in 'Verify Head' if a commit is made after the label is added but before the job runs 😄

**Review by dklilley — APPROVED** [review]

**Pull request merged**